### DISCUSSION TITLE
Better support of the return early pattern

### DISCUSSION THREAD

**Discussion by Khaos66** [dotnet/csharplang#7509] [Language Ideas] [answered]
I think it could be beneficial if `return` and others would support a `when` clause. This could supercharge the `return early pattern` and reduce indentation levels.

For example: Each event handler looks something like this right now:
```c#
void OnSomethinkChanged(object sender, EventArgs e)
{
   if (sender is SomeClass x)
   {
      x.DoSomething();
   }
}
```

If I would apply the return early pattern this would become:
```c#
void OnSomethinkChanged(object sender, EventArgs e)
{
   if (sender is not SomeClass x)
   {
      return;
   }

   x.DoSomething();
}
```

While this isn't bad, it could use some syntactic suggar ;)
```c#
void OnSomethinkChanged(object sender, EventArgs e)
{
   return when sender is not SomeClass x;
   x.DoSomething();
}
```

The same could be usefull with `continue` or `break` in loops. Also don't forget `yield`...

What do you think?

**Comment by HaloFour** [comment, marked as answer]
See: #2195

```cs
return when sender is not SomeClass x;

if (sender is not SomeClass x) return;
```

**Reply by Khaos66** [reply]
Yes. I get this. But when code style guidelines dictate that you use `{}` brackets, you're out of luck and back to 4 lines of if...

**Reply by HaloFour** [reply]
Those guidelines exist generally to avoid the issues that can arise with matching `else` statements.  But IMO in the case of early returns or other control statements it's pretty clear what the code is going to do so I feel that it's perfectly acceptable to not require a code block.  It's not like you can have additional statements following a `return`, `throw`, `continue` etc., that you intended to be in the conditional code block, and it's not all that common to use an `else` clause.

**Comment by Khaos66** [comment]
Ok. Thank you for pointing me to #2195. I've read this and the linked threads and agree (somewhat). 
I just thought `if (sender is not SomeClass x) { return; }` looks ugly. But I'll adjust ;)
